commands/bnet.js

```diff
@@ -0,0 +1,443 @@
+          printBattleTags(message.client, channelId);
+        }
+      }
+    }
+  });
+};
+
+/**
+ * Prints all of the currently stored BattleTags in the appropriate channel.
+ * Prefers editing exiting message to adding new ones, and removes messages by
+ * other users in the channel.
+ * @param {Discord.Client} client The Client to use to print
+ * @param {string} channelId The Snowflake ID of the channel to print to
+ */
+function printBattleTags(client, channelId) {
+  if (!battleTags) return;
+
+  // Build a list of messages to write.
+  const textMessages = [kHeaderString];
+  for (entry of [...battleTags.entries()].map((e) => e[1])
+      .sort(BattleTagEntry.compare)) {
+    const line = entry.toEntryString();
+    if (textMessages[textMessages.length - 1].length +
+        line.length + 1 > kCharsPerMessage) {
+      textMessages.push(line);
+    } else {
+      textMessages[textMessages.length - 1] += '\n' + line;
+    }
+  }
+
+  // Get the current channel messages.
+  const channel = getChannelById(client, channelId);
+  if (!channel) return;
+  channel.fetchMessages().then((messages) => {
+    // Filter to self messages.
+    const sortedMessages = [...messages.entries()].sort().map((e) => e[1]);
+    const selfMessages = sortedMessages.filter(
+        (m) => m.author.id === client.user.id);
+
+    // Update the contents of the existing messages.
+    for (i = 0; i < Math.min(selfMessages.length, textMessages.length); i++) {
+      selfMessages[i].edit(textMessages[i]);
+    }
+
+    // Add messages as necessary.
+    for (i = selfMessages.length; i < textMessages.length; i++) {
+      channel.send(textMessages[i]);
+    }
+
+    // Delete extra self messages if necessary.
+    for (i = textMessages.length; i < selfMessages.length; i++) {
+      selfMessages[i].delete();
+    }
+
+    // Delete all other messages in the channel.
+    for (message of sortedMessages.filter((m) =>
+      m.author.id !== client.user.id)) {
+      if (message.author.id !== client.user.id) {
+        message.delete();
+      }
+    }
+  });
+};
+
+/**
+ * Parses a possible mention string into a Snowflake string.
+ * @param {string} text The text to parse
+ * @return {string} The extracted Snowflake, or null if the text is not a
+ * mention.
+ */
+function parseSnowflake(text) {
+  const userSnowflake = text.match(kSnowflakeRegex);
+  return userSnowflake ? userSnowflake[1] : null;
+}
+
+/**
+ * Parses a possible mention string into a Snowflake string.
+ * @param {string} text The text to parse
+ * @return {string} The extracted Snowflake, or the input if the text is not a
+ * mention.
+ */
+function tryParseSnowflake(text) {
+  const userSnowflake = text.match(kSnowflakeRegex);
+  return userSnowflake ? userSnowflake[1] : text;
+};
+
+/**
+ * Checks the permissions of a user against the required roles.
+ * TODO(aalberg): Refactor this to a utility module.
+ * @param {Discord.User} user The user to check
```

Comment:
This is a GuildMember not a User (users apparently don't have a roles property).

Reply:
Done.